From f9b1033423530325198c9c31ab16d6dbf81c7e90 Mon Sep 17 00:00:00 2001
From: Ben Wolsieffer <benwolsieffer@gmail.com>
Date: Sat, 27 Aug 2022 00:56:11 -0400
Subject: [PATCH 3/3] ASoC: wm8960: use sysclk as MCLK if PLL is not configured

Simple-Card calls snd_soc_dai_set_sysclk() to configure the MCLK frequency, but
doesn't call snd_soc_dai_set_pll(). To support this case, fallback to using
the configured sysclk frequency as MCLK if the PLL has not been configured.

Signed-off-by: Ben Wolsieffer <benwolsieffer@gmail.com>
---
 sound/soc/codecs/wm8960.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/sound/soc/codecs/wm8960.c b/sound/soc/codecs/wm8960.c
index baaa8d2a3781..234c2ad0b307 100644
--- a/sound/soc/codecs/wm8960.c
+++ b/sound/soc/codecs/wm8960.c
@@ -761,12 +761,22 @@ static int wm8960_configure_clocking(struct snd_soc_component *component)
 		return 0;
 	}
 
-	if (wm8960->clk_id != WM8960_SYSCLK_SRC_MCLK && !wm8960->freq_in) {
+	/*
+	 * If a user has called set_pll(), use the configured input frequency as
+	 * MCLK. If not, try to use the frequency passed to set_sysclk() as
+	 * MCLK.
+	 */
+	if (wm8960->freq_in) {
+		freq_in = wm8960->freq_in;
+	} else {
+		freq_in = wm8960->sysclk;
+	}
+
+	if (wm8960->clk_id != WM8960_SYSCLK_SRC_MCLK && !freq_in) {
 		dev_err(component->dev, "No MCLK configured\n");
 		return -EINVAL;
 	}
 
-	freq_in = wm8960->freq_in;
 	/*
 	 * If it's sysclk auto mode, check if the MCLK can provide sysclk or
 	 * not. If MCLK can provide sysclk, using MCLK to provide sysclk
-- 
2.40.0

